cpp/src/io/avro/reader_impl.cu

```diff
@@ -146,104 +151,187 @@ class metadata : public file_metadata {
+          if (inflate_out[i].status == 1 && inflate_out[i].bytes_written > inflate_in[i].dstSize) {
+            inflate_in[i].dstSize = inflate_out[i].bytes_written;
+          }
+          actual_uncompressed_size += inflate_in[i].dstSize;
         }
-        actual_uncompressed_size += inflate_in[i].dstSize;
-      }
-      if (actual_uncompressed_size > uncompressed_data_size) {
-        decomp_block_data.resize(actual_uncompressed_size, stream);
-        for (size_t i = 0, dst_pos = 0; i < meta.block_list.size(); i++) {
-          auto dst_base           = static_cast<uint8_t*>(decomp_block_data.data());
-          inflate_in[i].dstDevice = dst_base + dst_pos;
-
-          meta.block_list[i].offset = dst_pos;
-          meta.block_list[i].size   = static_cast<uint32_t>(inflate_in[i].dstSize);
-          dst_pos += meta.block_list[i].size;
+        if (actual_uncompressed_size > uncompressed_data_size) {
+          decomp_block_data.resize(actual_uncompressed_size, stream);
+          for (size_t i = 0, dst_pos = 0; i < meta.block_list.size(); i++) {
+            auto dst_base           = static_cast<uint8_t*>(decomp_block_data.data());
+            inflate_in[i].dstDevice = dst_base + dst_pos;
+
+            meta.block_list[i].offset = dst_pos;
+            meta.block_list[i].size   = static_cast<uint32_t>(inflate_in[i].dstSize);
+            dst_pos += meta.block_list[i].size;
+          }
+        } else {
+          break;
         }
-      } else {
-        break;
       }
-    } else {
-      break;
     }
-  }
 
-  return decomp_block_data;
+    return decomp_block_data;
+  } else if (meta.codec == "snappy") {
+    size_t num_blocks = meta.block_list.size();
+
+    // comp_block_data contains contents of the avro file starting from the first block, excluding
+    // file header. meta.block_list[i].offset refers to offset of block i in the file, including
+    // file header.
+    // Find ptrs to each compressed block in comp_block_data by removing header offset.
+    hostdevice_vector<void const*> compressed_data_ptrs(num_blocks, stream);
+    std::transform(meta.block_list.begin(),
+                   meta.block_list.end(),
+                   compressed_data_ptrs.host_ptr(),
+                   [&](auto const& block) {
+                     return static_cast<std::byte const*>(comp_block_data.data()) +
+                            (block.offset - meta.block_list[0].offset);
+                   });
+    compressed_data_ptrs.host_to_device(stream);
+
+    hostdevice_vector<size_t> compressed_data_sizes(num_blocks, stream);
+    std::transform(meta.block_list.begin(),
+                   meta.block_list.end(),
+                   compressed_data_sizes.host_ptr(),
+                   [](auto const& block) { return block.size; });
+    compressed_data_sizes.host_to_device(stream);
+
+    hostdevice_vector<size_t> uncompressed_data_sizes(num_blocks, stream);
+    CUDF_EXPECTS(nvcompStatus_t::nvcompSuccess ==
```

Comment:
Might be worth separating out the call to `nvcompBatchedSnappGetDecompressSizeAsync()` and do the `CUDF_EXPECTS` separately.

Reply:
I changed that recently in the other readers as per [this review](https://github.com/rapidsai/cudf/pull/8252#discussion_r704941425). Didn't update this PR yet.